### ISSUE TITLE
Weapons used by monsters are not listed in their weapons list

### ISSUE THREAD

**Issue by pingurus** [canismarko/dungeon-sheets#111]
I use dungeonsheets as a base to my dungeonstorybook project and noticed that the weapons wielded by monsters are listed in their docstrings and not in their weapon list. The spells also are double-listed, while missing the `spell_slots` (See for example the lich).

Is this intentional? In my use-case it would be useful if the weapons and skills would be listed just as they would be for a character, and I assume that this would be generally useful.

Maybe @stravajiaxen could improve their script from #97 to fix this, otherwise I would be willing to try to fix this. Maybe some of the methods from `Character` would also have to be pulled up to `Creature` to do this cleanly.

**Comment by canismarko** [comment]
Glad to hear it's useful. You must be outputting to epub then, right? AFAIK, the duplicated spell list is only there in the epub output.

The first spell list comes from the "spellcasting" description which has spell slots and mirrors that found in the monster manual stat blocks. I added the second spell list so that I can make them links to the spell descriptions for quick reference when running a session. I thought about doing a search and replace for the name of each spell and then putting in <a> tags so that there's only one list, but it still contains links. The downside is that some spells that are also common words (e.g. "Shield") would be a bit overzealous in what they turned into links. Probably not a big deal though. That would solve your first problem, right?

I'm not sure monsters really should have a *weapons* attribute. It's inherited from the *Creature* class, but maybe that belongs on the *Character* class. Looking at a stat block, and the API at https://www.dnd5eapi.co, there doesn't seem to be a difference between a weapon attack and a spell/feature/etc. They're all just "actions". I think trying to parse out which actions are weapon attacks, only to then combine them again when making the output is probably not something I'm interested in.

I'm not really sure I understand what the problem is with monster weapons the way they are now. Could you give some more info on what you'd like to be different?

**Comment by pingurus** [comment]
> Glad to hear it's useful. You must be outputting to epub then, right? AFAIK, the duplicated spell list is only there in the epub output.

No, I output it into html using my own rst generator. I do it that way so I can comfortably link from other hand-written rst files to the automatically generated files. I may be duplicating code by doing that though.

> The first spell list comes from the "spellcasting" description which has spell slots and mirrors that found in the monster manual stat blocks. I added the second spell list so that I can make them links to the spell descriptions for quick reference when running a session. I thought about doing a search and replace for the name of each spell and then putting in tags so that there's only one list, but it still contains links. The downside is that some spells that are also common words (e.g. "Shield") would be a bit overzealous in what they turned into links. Probably not a big deal though. That would solve your first problem, right?

If I understand your intention correctly, yes. (I assume your first spell list is the one in the docstring and the second one is the one in the "spells" attribute, which I want to use)

> I'm not sure monsters really should have a weapons attribute. It's inherited from the Creature class, but maybe that belongs on the Character class. Looking at a stat block, and the API at https://www.dnd5eapi.co, there doesn't seem to be a difference between a weapon attack and a spell/feature/etc. They're all just "actions". I think trying to parse out which actions are weapon attacks, only to then combine them again when making the output is probably not something I'm interested in.

> I'm not really sure I understand what the problem is with monster weapons the way they are now. Could you give some more info on what you'd like to be different?

This is my current layout for a character:

![grafik](https://user-images.githubusercontent.com/36167579/132301767-852e066b-3d35-4a07-bcb7-a4166e93dd47.png)

And for skeletons for example this looks like this:

![grafik](https://user-images.githubusercontent.com/36167579/132301909-7b04efe8-4cff-4a74-80dd-31fbe34a457d.png)

However, I would want the weapons of a skeleton be listed as an attribute instead of in the docstring, so I can create the same table as for characters. I understand that we still have the actions which are specific to a monster which are unhandled then (claws, a tail etc.), but I would be generally in favour of havin the least needed information in the docstring so the representation of a class is more flexible. Also, I think there was some project that wants to help generate "fair" encounters, which needs well parsible attack actions, which is easier if they are listed in an attribute instead of text in a docstring.

Generally speaking I would prefer if characters and monsters would share the largest possible parts of their interface, as monsters in the end are only non-playable-characters with some treats that only those specific monsters have.

I hope I could clarify my issues :smile:
